receiver/prometheusoperatorreceiver/DESIGN.md

```diff
@@ -0,0 +1,142 @@
+
+In case of an [Agent](#Collector vs Agent deployment) deployment, which is signaled with the `limit_to_node` option, 
+only local endpoints will be fetched, endpoints should be filtered so that only pods are scraped which are scheduled 
+on the current node.
+To achieve this, additional relabeling rules are being added to all scrape configs generated by the `ConfigGenerator`
+
+Here an example of a scrape config with the additional rules:
+```yaml
+scrape_configs:
+- job_name: serviceMonitor/ingress/ingress-nginx-controller/0
+  honor_timestamps: true
+  scrape_interval: 30s
+  scrape_timeout: 10s
+  metrics_path: /metrics
+  scheme: http
+  follow_redirects: true
+  relabel_configs: # shortened
+### the autogenerated rules have been removed 
+    # save in case node ports
+  - source_labels: [__meta_kubernetes_endpoint_address_target_kind, __meta_kubernetes_endpoint_address_target_name]
+    regex: "Node;(.+)"
+    replacement: $1
+    target_label: __tmp_node_name
+    action: replace
+    # save in case pods
+  - source_labels: [__meta_kubernetes_endpoint_address_target_kind, __meta_kubernetes_endpoint_node_name]
+    regex: "Pod;(.+)"
+    replacement: $1
+    target_label: __tmp_node_name
+    action: replace
+    # keep only targets on the defined node
+  - source_labels: [__tmp_node_name]
+    regex: "our-node-name" # node name we have extracted from the environment
+    action: keep
+  kubernetes_sd_configs:
+  - role: endpoints
+    kubeconfig_file: ""
+    follow_redirects: true
+    namespaces:
+      names:
+      - ingress
+```
+
+This config is then used to create a `prometheusreceiver` `Config` struct.  
+
+### Major components of the config generation
+- **[ConfigGenerator](https://github.com/prometheus-operator/prometheus-operator/blob/main/pkg/prometheus/promcfg.go#L304):**
+  PrometheusOperator component which generates a valid Prometheus config marshalled to a byte array
+- **[ConfigLoader](https://github.com/prometheus/prometheus/blob/main/config/config.go#L68):**
+  Prometheus configuration loader which unmarshalls the config to a `prometheusreceiver` usable object 
+
+
+## Processing config change events
+The Prometheus config is at first compared against the currently applied one. Should there be any change, a new 
+`prometheusreceiver` is started using the generated configuration. 
+If the startup is successful the old instance of `prometheusreceiver` is shutdown. 
+Should any error occur during startup the old instance will keep running
+
+## Collector vs Agent deployment
+The receiver should support both reference architectures of the collector. 
+
+### Collector
+If running as collector the Prometheus config provided by PrometheusOperator can be reused without a change.    
+Should multiple instances with the same config run in the same cluster, they will act like a 
+high availability pair of Prometheus. Therefore, all targets will be scraped multiple times and telemetry 
```

Comment:
I haven't been aware of this till now. This is definitely a useful addition when the receiver is setup as collector!      
I will work this into the proposal

Reply:
The other thing to consider is the OpenTelemetry Operator's prometheus target allocation capability.  It is designed to allow running multiple collector instances and distributing targets across them.  It will re-allocate targets if a collector instance is added or removed.  I think adding the ability to utilize the pod and service monitors there should be considered as an alternative to building this into a receiver.

Reply:
@Aneurysm9 do you have a link to a design document for the target allocator? I couldn't find any in the OpentelemetryOperator repo or on opentelemetry.io.    

If the community prefers to implement this first in the target allocator, I will work on that instead.

Reply:
Here's the [initial outline of the capability](https://docs.google.com/document/d/1YRFIZl-N0Igr1uQQ_MVm10JX2-9hy9OItNDU2PJ-eiE/edit) and here's the [more detailed design doc for the target allocator service](https://docs.google.com/document/d/1GdpCqdkz72291l_8S-jio7EhVltminobvqshsFih6vA/edit#).

The target allocation server is set up to [reload its server when the config file it uses changes](https://github.com/open-telemetry/opentelemetry-operator/blob/main/cmd/otel-allocator/main.go#L83).  It should be feasible to add a watch for the Prometheus CRDs and use them to update the config, which will then cause the allocator to start using the generated SD configs.